### DISCUSSION TITLE
Imputing missing coordinates of a polymer

### DISCUSSION THREAD

**Discussion by spouyobo** [e3nn/e3nn#396] [Project wanted]
Thank you for this amazing framework! I stumbled across `e3nn` via [Andrew White's applied textbook on "Deep learning for molecules & materials"](https://dmol.pub/index.html). I'm enchanted by its possibility but lack the requisite background to fully understand the irreducible representations. I'm seeking helpful pointers to get me started with the model irreps. 

I'm studying the behavior of long polymers, and measuring their spatial position at specific monomers in the chain. Some of the positions of monomers are missing and I wish to impute them. For instance, 
```
polymer
>>>array([[        nan,         nan,         nan,  0.        ],
       [26.11893272,  6.39596176,  0.        ,  1.        ],
       [25.65973854,  6.70887423,  0.        ,  1.        ],
       [25.43806076,  7.01624107,  0.        ,  1.        ],
       [25.72443008,  7.21862507,  0.        ,  1.        ],
       [        nan,         nan,         nan,  0.        ],
       [26.21587372,  6.8746376 ,  0.        ,  1.        ],
       [        nan,         nan,         nan,  0.        ],
       [26.94106674,  6.65956831,  0.        ,  1.        ],
       [26.99314117,  6.26608562,  0.        ,  1.        ],
       [27.36512375,  6.04643583,  0.        ,  1.        ],
       [27.05812454,  5.77885437,  0.        ,  1.        ]])
```
is one such polymer where 12 positions are measured, with 3 positions missing. The columns correspond to `[x, y, z, mask]` where `mask == 0` if the position is missing and `1` otherwise.

I also happen to know the "molecular distance" between the positions.  For example,
```
mdist_mat
>>>array([[0., 2., 1., 2., 4., 0., 0., 1., 0., 5., 4., 2.],
       [2., 0., 1., 0., 2., 2., 2., 1., 2., 3., 2., 0.],
       [1., 1., 0., 1., 3., 1., 1., 0., 1., 4., 3., 1.],
       [2., 0., 1., 0., 2., 2., 2., 1., 2., 3., 2., 0.],
       [4., 2., 3., 2., 0., 4., 4., 3., 4., 1., 0., 2.],
       [0., 2., 1., 2., 4., 0., 0., 1., 0., 5., 4., 2.],
       [0., 2., 1., 2., 4., 0., 0., 1., 0., 5., 4., 2.],
       [1., 1., 0., 1., 3., 1., 1., 0., 1., 4., 3., 1.],
       [0., 2., 1., 2., 4., 0., 0., 1., 0., 5., 4., 2.],
       [5., 3., 4., 3., 1., 5., 5., 4., 5., 0., 1., 3.],
       [4., 2., 3., 2., 0., 4., 4., 3., 4., 1., 0., 2.],
       [2., 0., 1., 0., 2., 2., 2., 1., 2., 3., 2., 0.]])
```
represents how many monomers separate each of my measured locations. My intuition tells me this "molecular separation" should be related to the "spatial separation", and can inform my prediction of missing locations. I wish to utilize these as edge weights between my nodes. 

My questions are:
1. If I have a PyG dataset where I plan to label my **node features (`data.x`)** with my `mask` column (`0` to tell the model which indeces to impute), what is an appropriate `irreps_in`? What is the difference between `irreps_node_attr`, and `irreps_in`?
2. If I have a PyG dataset where I plan to label my **edge features (`data.edge_attr`)** with my `mdist_mat`, is it possible to still use this information somehow? Reading the documentation it seems `irreps_edge_attr` is usually designated with `e3nn.o3.Irreps.spherical_harmonics(3)`, and I don't see anywhere else I can pass edge attributes.

Here is the example model I am considering, based on the textbook:
```
model_kwargs = {
    "irreps_in": None,  # ???
    "irreps_hidden": e3nn.o3.Irreps("5x0e + 5x0o + 5x1e + 5x1o"),  
    "irreps_out": "1x1o",  
    "irreps_node_attr": None, # ???
    "irreps_edge_attr": e3nn.o3.Irreps.spherical_harmonics(3), # ???
    "layers": 3, 
    "max_radius": 3.5,
    "number_of_basis": 10,
    "radial_layers": 1,
    "radial_neurons": 128,
    "num_neighbors": 11, 
    "num_nodes": 12,  
    "reduce_output": False,  
}

model = e3nn.nn.models.gate_points_2101.Network(
    **model_kwargs
)  
```

**Comment by blondegeek** [comment]
I think it would be better to put a "dummy position" that then get's updated by context (operations later in the network). There really isn't a proper way to handle a feature that is not embedded in space unless it's a global variable. Hope that helps a bit.

**Reply by spouyobo** [reply]
Thanks for this! I linearly interpolated a value as a "placeholder", but forgot to mention this. I then pass in a column of node features to indicate which nodes have "placeholder" values, setting `irreps_node_attr` to `1x0e`, but I am not sure if this is the best approach.

Because I feel this task should be learnable, I am confused on how to best take advantage of my edge attributes - namely the molecular distance. Is it fair to say by setting `irreps_edge_attr` → `e3nn.o3.Irreps.spherical_harmonics(3)`, my edge attributes aren't used as features during training? I'm hoping there is a way to specify that the relationship between the edge attributes and the node positions are also Euclidean.

**Reply by mariogeiger** [reply]
By setting `irreps_edge_attr → e3nn.o3.Irreps.spherical_harmonics(3)` the model will use as edge attribute the spherical harmonics `0e+1o+2e` evaluated at the relative positions between the nodes.

**Reply by spouyobo** [reply]
Thanks for this confirmation! I'm grateful for the opportunity to start a discussion. I'll think about other strategies to get my `e3nn` model to learn to do this task.
